File: test/spec/simple_beam.js
/*jshint undef: true, unused: true */
/*global __dirname describe it require*/
var ROOT = __dirname + '/../..', SRC = ROOT + '/src';
var expect = require('expect.js');
var fe = require(SRC);

describe('FAESOR simple_beam example', function() {

  it('should create model.', function() {

    var FeNodeSet = fe.fens.FeNodeSet;
    var H8Block = fe.mesh.H8Block;
    var LinElIso = fe.property.LinElIso;
    var DeforSSLinElTriax = fe.material.DeforSSLinElTriax;
    var GaussRule = fe.integrationrule.GaussRule;
    var SparseSystemMatrix = fe.system.matrix.SparseSystemMatrix;
    var SparseSystemVector = fe.system.vector.SparseSystemVector;
    var mldivide = fe.system.mldivide;
    var Field = fe.field.Field;
    var EBC = fe.ebc.EBC;
    var DeforSS = fe.feblock.DeforSS;
    var NodalLoad = fe.nodalload.NodalLoad;
    var ForceIntensity = fe.forceintensity.ForceIntensity;
    var genISORm = fe.feutils.genISORm;
    var matrixEquals = fe.numeric.matrixEquals;
    var sum = fe.numeric.sum;
    var nthColumn = fe.numeric.nthColumn;

    // parameters:
    var E = 1000;
    var nu = 0.0;
    var W = 2.5;
    var H = 5;
    var L = 50;
    var htol = Math.min(L, H, W)/1000;
    var uzex = -12.6;
    var magn = 0.2*uzex/4;
    var Force = magn*W*H*2;

    var nx = 3;
    var mult = 4;
    var ny = mult*nx;
    var nz = 2*nx;

    var feb, febIr, surfaceIr, fens, gcells, mater, prop, ebcs, geom, u;

    var mesh = H8Block(W, L, H, nx, ny, nz);
    fens = mesh.fens();
    gcells = mesh.gcells();
    // console.log("fens = ", fens.xyz());
    // console.log("gcells = ", gcells);
    return;

    prop = new LinElIso({ E: E, nu: nu });

    mater = new DeforSSLinElTriax({
      property: prop
    });

    febIr = new GaussRule(3, 2);

    feb = new DeforSS({
      material: mater,
      gcells: gcells,
      integrationRule: febIr,
      rm: genISORm
    });

    geom = new Field({
      name: 'geom',
      fens: fens
    });

    ebcs = [
      {
        id: fens.boxSelect({
          bounds: [0, W, 0, 0, 0, H],
          inflate: 1e-4
        }),
        dir: [1, 2, 3],
        value: 0
      },
      {
        id: fens.boxSelect({
          bounds: [W, W, 0, L, 0, H],
          inflate: 1e-4
        }),
        dir: 1,
        value: 0
      }
    ].map(function(o) { return new EBC(o); });

    u = new Field({
      name: 'u',
      dim: geom.dim(),
      nfens: geom.nfens(),
      ebcs: ebcs
    });


    var elementMatrices = feb.stiffness(geom, u);
    var neqns = u.neqns();
    console.log("neqns = ", neqns);

    var K = new SparseSystemMatrix(neqns, neqns, elementMatrices);
    console.log("K = ", K.toFull());

    var fi = new ForceIntensity({ magn: [0, 0, magn] });
    var bdryGcells = gcells.boundary();
    var bcl = bdryGcells.boxSelect(fens, {
      bounds: [0, W, L, L, 0, H],
      inflate: htol
    });
    var lfeb = new DeforSS({
      mater: mater,
      gcells: bdryGcells.subset(bcl),
      integrationRule: surfaceIr
    });

    var elementVectors = lfeb.distributeLoads(geom, u, fi, 2);
    console.log("elementVectors = ", elementVectors);

    var F = new SparseSystemVector(neqns, elementVectors);
    console.log("F = ", F.toFull());

    var x = mldivide(K, F);
    console.log("x = ", x);

    u.scatterSystemVector_(x);
    var values = u.values();
    console.log("values = ", values);

    var freeendNids = fens.boxSelect({
      bounds: [0, W, L, L, 0, H],
      inflate: 1e-4
    });

    var uv = u.gatherValuesMatrix(freeendNids);
    var uz = sum(nthColumn(uv, 3))/freeendNids.length;
    console.log("uz = ", uz);

    var uzExpected = -9.4;
    console.log("uzExpected = ", uzExpected);

    expect(Math.abs(uz - uzExpected) < 1e-4).to.be(true);
    return 0;
  });

});

File: tools/faesor/simple_beam.m
function  simple_beam(n)
  E=1000;
  nu=0.4999999;
  nu=0.;
  W=2.5;
  H=5;
  L= 50;
  htol=min([L,H,W])/1000;
  uzex =-12.6;
  magn = 0.2*uzex/4;
  Force =magn*W*H*2;
  Force*L^3/(3*E*W*H^3*2/12);
  graphics = true;

  ## Number of elements
  #n=12;
  mult =4;

  eix=1;

  ## Selective reduced integration hexahedron
  ## eltyd(eix).mf =@H8_block;
  ## eltyd(eix).blf =@feblock_defor_ss_sri;
  ## eltyd(eix).integration_rule=gauss_rule (3,[1,2]);
  ## eltyd(eix).surface_integration_rule=gauss_rule(2, 2);
  ## eltyd(eix).styl='md-';
  ## eix=eix+1;

  %% Full integration hexahedron
  eltyd(eix).mf =@H8_block;
  eltyd(eix).blf =@feblock_defor_ss;
  eltyd(eix).integration_rule=gauss_rule (3,2);
  eltyd(eix).surface_integration_rule=gauss_rule(2, 2);
  eltyd(eix).styl='mx--';
  eix=eix+1;

  % Reduced integration H20
  ## eltyd(eix).mf =@H20_block;
  ## eltyd(eix).blf =@feblock_defor_ss;
  ## eltyd(eix).integration_rule=gauss_rule (3,2);
  ## eltyd(eix).surface_integration_rule=gauss_rule(2, 2);
  ## eltyd(eix).styl='ro-';
  ## eix=eix+1;

  ## Full integration H20
  ## eltyd(eix).mf =@H20_block;
  ## eltyd(eix).blf =@feblock_defor_ss;
  ## eltyd(eix).integration_rule=gauss_rule (3,3);
  ## eltyd(eix).surface_integration_rule=gauss_rule(2, 2);
  ## eltyd(eix).styl='b*-';
  ## eix=eix+1;

  ## T10 tetrahedron
  ## eltyd(eix).mf =@T10_block;
  ## eltyd(eix).blf =@feblock_defor_ss;
  ## eltyd(eix).integration_rule=tet_rule (4);
  ## eltyd(eix).surface_integration_rule=tri_rule(3);
  ## eltyd(eix).styl='gd-';
  ## eix=eix+1;

  eix = 1;

  ## Create the mesh and initialize the geometry
  [fens,gcells]= feval (eltyd(eix).mf, W, L, H, n,mult*n,2*n);

  xyz = get(fens, 'xyz');
  xyz(101:105, :)

  conn = get(gcells, 'conn');
  conn(101:105, :)

  ## mesh{1}=fens;
  ## mesh{2}=gcells;
  ## drawmesh(mesh,'shrink', 0.9,'facecolor','red'); view (2); pause
  ## return;

  prop = property_linel_iso (struct('E',E,'nu',nu));
  mater = mater_defor_ss_linel_triax (struct('property',prop));

  feb = feval (eltyd(eix).blf, struct ('mater',mater, 'gcells',gcells, 'integration_rule',eltyd(eix).integration_rule));
  geom = field(struct ('name',['geom'], 'dim', 3, 'fens',fens));

  ## Define the displacement field, and zero it out
  u   = 0*geom;

  %% Apply the EBC''s
  ebc_fenids=fenode_select(fens,struct('box',[0 W 0 0 0 H]));
  ebc_prescribed=ones(length(ebc_fenids),1);
  ebc_comp=[];
  ebc_val=zeros(length(ebc_fenids),1);
  u   = set_ebc(u, ebc_fenids, ebc_prescribed, ebc_comp, ebc_val);
  ebc_fenids=fenode_select(fens,struct('box',[W W 0 L 0 H]));
  ebc_prescribed=ones(length(ebc_fenids),1);
  ebc_comp=ebc_fenids*0+1;
  ebc_val=zeros(length(ebc_fenids),1);
  u   = set_ebc(u, ebc_fenids, ebc_prescribed, ebc_comp, ebc_val);
  u   = apply_ebc (u);

  %% Number the equations
  u   = numbereqns (u);
  disp (['n=' num2str(n) ', number of equations =' num2str(get(u,'neqns'))])

  %% Assemble the stiffness matrix
  %% K = start (sparse_iter_sysmat, get(u, 'neqns'));
  K = start (sparse_sysmat, get(u, 'neqns'));
                    % K = start (sparse_chol_sysmat, get(u, 'neqns'));
  K = assemble (K, stiffness(feb, geom, u));
  ## mat = get(K, 'mat');
  ## fprintf('k:\n')
  ## full(mat([32,40,7,18], [32,40,7,18]))
  ## mat(3,3)
  ## mat(4,4)
  ## mat(77,77)
  ## mat(8,8)

  %% Assemble the load vector
  fi= force_intensity(struct ('magn',[0;0; magn]));
  bdry_gcells = mesh_bdry(gcells, []);

  nbdrycells = count(bdry_gcells);
  bdry_conn = get(bdry_gcells, 'conn');
  ## bdry_conn = normalizedConn(bdry_conn)
  ## bdry_conn_201_205 = bdry_conn(201:205, :)

  bcl = gcell_select(fens, bdry_gcells, ...
                     struct ('box',[0 W L L 0 H],'inflate',htol));

  selected_bdry_cells = subset(bdry_gcells,bcl);
  selected_bdry_cells_conn = get(selected_bdry_cells, 'conn');
  ## selected_bdry_cells_conn = normalizedConn(selected_bdry_cells_conn)

  lfeb = feblock_defor_ss(struct ('mater',mater, 'gcells',subset(bdry_gcells,bcl),...
                                  'integration_rule',eltyd(eix).surface_integration_rule));
  F = start (sysvec, get(u, 'neqns'));

  evs = distrib_loads(lfeb, geom, u, fi,2);
  ## evs_vec = get(evs, 'vec');
  ## evs_eqnums = get(evs, 'eqnums');
  ## for i=1:length(evs_vec)
  ##   i
  ##   evs_vec{i}
  ##   evs_eqnums{i}
  ## end
  F = assemble (F, distrib_loads(lfeb, geom, u, fi,2));


  %% Solve K*u=F
  b=get(F,'vec');
  ## tic;
  x=K\b;
  ## toc

  %% Transfer the solution to the field u
  u = scatter_sysvec(u, x);
  ebc_fenids=fenode_select(fens,struct('box',[0 W L L 0 H]));
  uv=gather (u,ebc_fenids,'values','noreshape');
  uz=sum(uv(:, 3))/length(ebc_fenids);
  disp (['   Free end displacement =' num2str(uz)])

end
